doc/barclamp_development_exercises.md

```diff
@@ -0,0 +1,341 @@
+upgrade migration needs to take the new parameter from the template attributes
+hash and add it to the attributes hash, and the downgrade needs to do the
+opposite. It will look something like this:
+
+```
+def upgrade(template_attrs, template_deployment, attrs, deployment)
+  attrs["insecure_debug"] = template_attrs["insecure_debug"] unless attrs.key?("insecure_debug")
+  return attrs, deployment
+end
+
+def downgrade(template_attrs, template_deployment, attrs, deployment)
+  attrs.delete("insecure_debug") unless template_attrs.key?("insecure_debug")
+  return attrs, deployment
+end
+```
+
+Run the migration. There are two ways to do this:
+
+```
+# barclamp_install.rb /opt/dell/crowbar_framework/barclamps
+```
+
+will reinstall all the barclamps, including running all the migrations. Don't be
+confused by the ``.rb`` file suffix, the command should be an executable already
+in your PATH. Alternatively:
+
+```
+# cd /opt/dell/crowbar_framework
+# RAILS_ENV=production rake crowbar:schema_migrate
+# RAILS_ENV=production rake crowbar:schema_migrate_status # optional
+```
+
+You can use the built-in rake commands to do the migration.
+``crowbar:schema_migrate`` does the migration, ``crowbar:schema_migrate_status``
+shows the revision number for each barclamp. You can use ``rake -T`` while in a
+directory containing a ``Rakefile`` to see all the available commands. The
+default Rails environment is ``development`` but we never configure a
+development environment, so you need to specify ``production``.
+
+Once the migration is run, check the raw view of the barclamp in the crowbar UI
+to make sure the parameter appears.
+
+#### Add the parameter to the cookbook
+
+The cookbook contains the Ruby code that creates and applies changes to the
+deployment.
+
+Edit the keystone server recipe in
+``/opt/dell/chef/cookbooks/keystone/recipes/server.rb``. Find the template
+resource that creates the keystone config file and add the new parameter to the
+list of variables it passes to the template.
+
+Edit the keystone config template in
+``/opt/dell/chef/cookbooks/keystone/templates/default/keystone.conf.erb`` to use
+the new parameter.
+
+Chef does not pick up changes to cookbooks automatically, they need to be
+uploaded to the Chef server:
+
+```
+# knife cookbook upload keystone -o /opt/dell/chef/cookbooks
+```
+
+Now you should be able to apply the change to the deployment. In the crowbar UI,
+click "Apply". The new parameter should appear in the keystone config file after
+the proposal is finished running. If you change the value in the raw view of the
+proposal and reapply, the config file will have changed.
+
+#### Add the parameter to the UI
+
+In many cases this is as far as we go, but for some common operations it is good
+to expose the parameter in the "Custom" view of the crowbar UI.
+
+Edit the locales file at
+``/opt/dell/crowbar_framework/config/locales/keystone/en.yml``. Add the new
```

Comment:
path relative to git perhaps?

Reply:
I explained why I'm preferring the package-installed location in the intro